docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,155 @@
+# ADR 022: Custom baseapp panic handling
+
+## Changelog
+
+- 2020 Apr 24: Initial Draft
+
+## Context
+
+Current implementation of BaseApp does not allow developers to write custom error handlers in
+[runTx()](https://github.com/cosmos/cosmos-sdk/blob/bad4ca75f58b182f600396ca350ad844c18fc80b/baseapp/baseapp.go#L538)
+method. We think that this method can be more flexible and can give SDK users more options for customizations without
+the need to rewrite whole BaseApp. Also there's one special case for `sdk.ErrorOutOfGas` error which feels like dirty
+hack in non-flexible environment.
+
+We propose middleware-solution, which could help developers implement following cases:
+* add external logging (let's say sending reports to external services like Sentry);
+* call panic for specific error cases;
+
+It will also make `OutOfGas` case and `default` case one of the middlewares.
+
+## Decision
+
+### Design
+
+#### Overview
+
+Instead of hardcoding custom error handling into BaseApp we suggest using set of middlewares which can be customized
+externally and will allow developers use as many custom error handlers as they want. Implementation with tests
+can be found [here](https://github.com/cosmos/cosmos-sdk/pull/6053).
+
+#### Implementation details
+
+##### RecoveryHandler
+
+We add a `recover()` object handler type:
+
+```go
+type RecoveryHandler func(recoveryObj interface{}) error
+```
+
+This handler receives an object, processes it and returns contextually enriched `error` or `nil`
+if object is not a handle's target type.
+
+An example:
+
+```go
+func(recoveryObj interface{}) error {
+    err, ok := recoveryObj.(error)
+    if !ok { return nil }
+    
+    if someSpecificError.Is(err) {
+        panic(customPanicMsg)
+    } else {
+        return nil
+    }
+}
+```
+
+This example breaks the application execution, but it also might enrich the error's context like the `OutOfGas` handler.
+
+##### RecoveryMiddleware
+
+We also add a middleware type:
+
+```go
+type recoveryMiddleware func(recoveryObj interface{}) (recoveryMiddleware, error)
+```
+
+Function receives a `recover()` object and returns (`next middleware`, `nil`) if object wasn't handled (not a target type)
```

Comment:
since this looks like the modular anteHandler (ADR10) I think it deserves a mention for comparison. 👍